src/smtp_server.c

```diff
@@ -123,59 +139,82 @@ int smtp_server_response(bfd_t *f, int code, const char *message)
 int smtp_server_process(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream)
 {
 	int schs, continue_session = 1;
+	int hdlr_idx;
+	int code;
+
 	struct smtp_cmd_hdlr_list *hlink;
+	struct smtp_cmd_hdlr *cmd_hdlr;
+	char *message;
 
 	do {
-		/* Save ctx->node to local var, since it can be changed by a
-		 * command handler *while* we are walking the handler list */
-		struct smtp_cmd_tree *node = ctx->node;
-
-		ctx->code = 0;
-		ctx->message = ctx->prev_message = NULL;
+		code = 0;
+		message = NULL;
+	    hdlr_idx = smtp_get_hdlr_idx(cmd);
 
-		list_for_each_entry(hlink, &node->hdlrs, lh) {
+	    if (hdlr_idx != -1) {
+			/* Get the structure with specific handler */
+			cmd_hdlr = &smtp_cmd_hdlrs[hdlr_idx];
 
-			if (ctx->prev_message != NULL)
-				free(ctx->prev_message);
-			ctx->prev_code = ctx->code;
-			ctx->prev_message = ctx->message;
-			ctx->code = 0;
-			ctx->message = NULL;
+			/* Call the handler */
+			schs = cmd_hdlr->smtp_preprocess_hdlr(ctx, cmd, arg, stream);
```

Comment:
All "schs" logic made sense for the old system that called multiple handlers for each smtp command.

The logic in smtp_server_process() is very simple. The "new" smtp command handlers should return the "disconnect" value received from the JS call. The do/while loop should continue as long as a return value of 0 is received from the smtp command handler. For clarity, "continue_session" should be renamed to "disconnect" and the do/while loop should test for this value.

The code and message values should always be taken from ctx->code and ctx->message, respectively - except for the case when smtp_get_hdlr_idx() returns -1 (in that case it's ok to assign direct values like you do now).